File: dart/0347-top-k-frequent-elements.dart
class Solution {
  List<int> topKFrequent(List<int> nums, int k) {
    Map<int, int> frequencyMap = {};
    for (int num in nums) {
      frequencyMap[num] = (frequencyMap[num] ?? 0) + 1;
    }

    List<List<int>> frequencyList =
        List.filled(nums.length + 1, List<int>.empty(growable: true));

    frequencyMap.forEach((num, count) {
      frequencyList[count] = [...frequencyList[count], num];
    });

    List<int> res = [];

    for (var i = frequencyList.length - 1; i >= 0; i--) {
      res.addAll(frequencyList[i]);

      if (res.length == k) {
        break;
      }
    }

    return res;
  }
}

class Solution {
  List<int> topKFrequent(List<int> nums, int k) {
    Map<int, int> map = {};
    for (int n in nums) {
      map[n] = (map[n] ?? 0) + 1;
    }
    List<int> uniqueNums = map.keys.toList();
    uniqueNums.sort((a, b) => map[b]!.compareTo(map[a]!));
    return uniqueNums.sublist(0, k);
  }
}

File: csharp/0347-top-k-frequent-elements.cs
public class Solution
{
    public int[] TopKFrequent(int[] nums, int k)
    {
        Dictionary<int, int> numCount = new Dictionary<int, int>();
        for (int i = 0; i < nums.Length; i++)
        {
            if (numCount.ContainsKey(nums[i]))
                numCount[nums[i]] += 1;
            else
                numCount[nums[i]] = 1;
        }

        List<int>[] bucketSort = new List<int>[nums.Length + 1];

        for (int i = 0; i < bucketSort.Length; i++)
            bucketSort[i] = new List<int>();

        foreach (var i in numCount)
            bucketSort[i.Value].Add(i.Key);

        int[] result = new int[k];
        int resultCount = 0;

        for (int i = bucketSort.Length - 1; i >= 1; i--)
        {
            if (bucketSort[i].Count == 0)
                continue;

            foreach (int item in bucketSort[i])
            {
                result[resultCount] = item;
                resultCount++;
            }
            if (resultCount == k)
                break;
        }
        return result;
    }
}